File: app/src/main/java/com/example/smartspend/View/HistoryActivity.java
// app/src/main/java/com/example/smartspend/view/HistoryActivity.java
package com.example.smartspend.view;

import android.app.DatePickerDialog;
import android.os.Bundle;
import android.widget.Button;
import android.widget.SearchView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.smartspend.R;
import com.example.smartspend.adapter.TransactionAdapter;
import com.example.smartspend.viewmodel.HistoryViewModel;
import com.example.smartspend.viewmodel.HistoryViewModelFactory;

import java.util.Calendar;

public class HistoryActivity extends AppCompatActivity {

    private HistoryViewModel viewModel;
    private TransactionAdapter adapter;

    private long startDate = -1, endDate = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_history);

        // TODO : remplacer par l'userId réel depuis la session
        int userId = getSharedPreferences("session", MODE_PRIVATE).getInt("userId", -1);

        // ViewModel
        HistoryViewModelFactory factory = new HistoryViewModelFactory(getApplication(), userId);
        viewModel = new ViewModelProvider(this, factory).get(HistoryViewModel.class);

        // RecyclerView
        RecyclerView recyclerView = findViewById(R.id.recycler_history);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        adapter = new TransactionAdapter();
        recyclerView.setAdapter(adapter);

        // Observer LiveData
        viewModel.transactions.observe(this, adapter::setTransactions);

        // Clic sur une transaction → édition (à brancher sur le Membre 3)
        adapter.setOnTransactionClickListener(transaction ->
                Toast.makeText(this, "Transaction : " + transaction.category, Toast.LENGTH_SHORT).show()
        );

        // SearchView
        SearchView searchView = findViewById(R.id.search_view);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override public boolean onQueryTextSubmit(String query) { return false; }
            @Override public boolean onQueryTextChange(String newText) {
                viewModel.search(newText);
                return true;
            }
        });

        // Filtre par date
        Button btnStartDate = findViewById(R.id.btn_start_date);
        Button btnEndDate   = findViewById(R.id.btn_end_date);
        Button btnReset     = findViewById(R.id.btn_reset_filter);

        btnStartDate.setOnClickListener(v -> pickDate(true, btnStartDate));
        btnEndDate.setOnClickListener(v -> pickDate(false, btnEndDate));
        btnReset.setOnClickListener(v -> {
            startDate = -1; endDate = -1;
            btnStartDate.setText("Date début");
            btnEndDate.setText("Date fin");
            viewModel.resetFilter();
        });
    }

    private void pickDate(boolean isStart, Button btn) {
        Calendar cal = Calendar.getInstance();
        new DatePickerDialog(this, (view, year, month, day) -> {
            Calendar selected = Calendar.getInstance();
            selected.set(year, month, day, isStart ? 0 : 23, isStart ? 0 : 59, isStart ? 0 : 59);
            long timestamp = selected.getTimeInMillis();
            String label = day + "/" + (month + 1) + "/" + year;
            btn.setText(label);
            if (isStart) startDate = timestamp;
            else endDate = timestamp;

            if (startDate != -1 && endDate != -1)
                viewModel.filterByDateRange(startDate, endDate);

        }, cal.get(Calendar.YEAR), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH)).show();
    }
}

File: app/src/main/java/view/HistoryActivity.kt
package com.team.smartspend.view

import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.SearchView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.team.smartspend.R
import com.team.smartspend.model.Transaction
import com.team.smartspend.utils.ExportUtils
import com.team.smartspend.viewmodel.HistoryViewModel
import java.util.Calendar

/**
 * HistoryActivity — écran Historique (Membre 4, adapté en Kotlin).
 *
 * Fonctionnalités : liste (RecyclerView triée par date décroissante), recherche
 * temps réel, filtre par plage de dates, export CSV/PDF, et clic sur une
 * transaction pour la modifier (réutilise les écrans du Membre 3).
 */
class HistoryActivity : AppCompatActivity() {

    private lateinit var viewModel: HistoryViewModel
    private lateinit var adapter: HistoryAdapter

    private var startDate: Long = -1
    private var endDate: Long = -1

    // Dernière liste affichée (sert à l'export)
    private var listeAffichee: List<Transaction> = emptyList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_history)

        viewModel = ViewModelProvider(this)[HistoryViewModel::class.java]

        // RecyclerView
        val recyclerView = findViewById<RecyclerView>(R.id.recycler_history)
        recyclerView.layoutManager = LinearLayoutManager(this)
        adapter = HistoryAdapter()
        recyclerView.adapter = adapter

        // Observer la liste (réactif)
        viewModel.transactions.observe(this) { liste ->
            listeAffichee = liste
            adapter.setTransactions(liste)
        }

        // Clic sur une transaction -> édition (écrans du Membre 3)
        adapter.setOnTransactionClickListener { t -> ouvrirEdition(t) }

        // Recherche en temps réel
        val searchView = findViewById<SearchView>(R.id.search_view)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean = false
            override fun onQueryTextChange(newText: String?): Boolean {
                viewModel.search(newText ?: "")
                return true
            }
        })

        // Filtre par date
        val btnStart = findViewById<Button>(R.id.btn_start_date)
        val btnEnd = findViewById<Button>(R.id.btn_end_date)
        val btnReset = findViewById<Button>(R.id.btn_reset_filter)

        btnStart.setOnClickListener { pickDate(true, btnStart) }
        btnEnd.setOnClickListener { pickDate(false, btnEnd) }
        btnReset.setOnClickListener {
            startDate = -1; endDate = -1
            btnStart.text = "Date début"
            btnEnd.text = "Date fin"
            searchView.setQuery("", false)
            viewModel.resetFilter()
        }

        // Export
        findViewById<Button>(R.id.btn_export_csv).setOnClickListener {
            val ok = ExportUtils.exportCSV(this, listeAffichee)
            Toast.makeText(this,
                if (ok) "✅ Export CSV réussi (dossier Téléchargements)" else "❌ Échec de l'export CSV",
                Toast.LENGTH_LONG).show()
        }
        findViewById<Button>(R.id.btn_export_pdf).setOnClickListener {
            val ok = ExportUtils.exportPDF(this, listeAffichee)
            Toast.makeText(this,
                if (ok) "✅ Export PDF réussi (dossier Téléchargements)" else "❌ Échec de l'export PDF",
                Toast.LENGTH_LONG).show()
        }
    }

    private fun pickDate(isStart: Boolean, btn: Button) {
        val cal = Calendar.getInstance()
        DatePickerDialog(this, { _, year, month, day ->
            val selected = Calendar.getInstance()
            // début de journée pour la date début, fin de journée pour la date fin
            if (isStart) selected.set(year, month, day, 0, 0, 0)
            else selected.set(year, month, day, 23, 59, 59)
            val timestamp = selected.timeInMillis
            btn.text = "$day/${month + 1}/$year"
            if (isStart) startDate = timestamp else endDate = timestamp

            if (startDate != -1L && endDate != -1L) {
                viewModel.filterByDateRange(startDate, endDate)
            }
        }, cal.get(Calendar.YEAR), cal.get(Calendar.MONTH), cal.get(Calendar.DAY_OF_MONTH)).show()
    }

    private fun ouvrirEdition(t: Transaction) {
        val intent = if (t.type == "REVENU") {
            Intent(this, AddIncomeActivity::class.java)
                .putExtra(AddIncomeActivity.EXTRA_ID, t.id)
        } else {
            Intent(this, AddExpenseActivity::class.java)
                .putExtra(AddExpenseActivity.EXTRA_ID, t.id)
        }
        startActivity(intent)
    }
}
